Feature/typings

SzybkiSasza opened the pull request:
As Typescript gets more and more popular, adding typings to the repository would make importing package in Typescript environments much easier.

I tried to keep it simple and concise and affect as little as possible, the only external dependency being Webpack @types/webpack (not Webpack itself).

SzybkiSasza commented:
The code isn't very complex and it's already written using ES6 classes - shouldn't be problematic, thanks to that ;) 

The biggest challenge would be to properly build so it supports both Import/export syntax and commonJS.

I might try helping with that in my free time, but it'd take a bit to standardize and properly test.

mattlewis92 commented:
> The biggest challenge would be to properly build so it supports both Import/export syntax and commonJS.

You don't need to worry about that, ES6 modules are only useful for frontend where tree shaking is a thing, for node you just need to transpile to commonjs and it will work everywhere

SzybkiSasza commented:
Perfect, I'll try to conjure sth in following days then ;)

SzybkiSasza commented:
@mattlewis92 Basic work is done. All the tests are passing and all the code (including tests) is rewritten to Typescript, with a few improvements :)

I'll need another day to refine the tests (cover new type guards) and probably address https://github.com/mattlewis92/webpack-filter-warnings-plugin/pull/2 as well.

SzybkiSasza commented:
@mattlewis92 I've just finished coding! :) Tested the library with ES imports - it works fine.

- Added function as an exclude option
- Refined type checking for `exclude`
- Modified `package.json` to support different TS checks
- Updated `security` script to use `npm audit`

And a lot of smaller changes. Feel free to merge and if you don't mind - I'd be thankful if you can mention me as a contributor / typescript version author ;)

SzybkiSasza commented:
One remark - I changed `main` to `index.js` which exports ES module. `cjs` is still available under `dist/cjs`. Do you think that's acceptable (would probably break for older users, hence - major version update would be needed)?

SzybkiSasza commented:
@mattlewis92 another round of changes - I was able to update code structure to prepare two builds:

- ES module - compatible that doesn't transpile exports
- CJS module that works in any NodeJS environment

I also added some build tools and tests for them. I tested both ES and CJS and typings for TS. Importing is not touched after that last update - main exported module is CJS.

I also updated README. Please feel free to adjust it in any desirable way ;)

SzybkiSasza commented:
Hi! I'm sorry for not responding for so long, but I was switching commercial projects and well... Didn't have any time for Open Source :)

- For Webpack - I completely agree - we can safely assume that someone using this plugin will add Webpack typings as well. I'll also update modified README to inform about that in Typescript section.
- About separate indexes - I remember I modelled it after one repo I found online (this solution guarantees proper separation between both), but I'm not so sure anymore - maybe I took one step too far. I'll experiment with it and update repo soon :)

SzybkiSasza commented:
I updated the PR:

- `@types/webpack` were moved to devDependencies and peerDependencies. That way, they won't be included in built files (as long as user is using NPM >= 3.x), but build itself will succeed.
- As for ES modules / CommonJS - I experimented with different ways of exporting the plugin and if you agree about CommonJS exporting `{WebpackFilterWarningsPlugin}` as an object, then we're set. I'm against default exports, as they're considered harmful / problematic, according to at least a few sources: https://basarat.gitbooks.io/typescript/docs/tips/defaultIsBad.html https://blog.neufund.org/why-we-have-banned-default-exports-and-you-should-do-the-same-d51fdc2cf2ad

So... If you like the changes and approve, we're set!

mattlewis92 commented on package.json:
```diff
@@ -2,23 +2,31 @@
+    "build:babel": "babel ts-out --out-dir . --copy-files ",
+    "build:cjs": "tsc -p './build-tools/tsconfig-cjs.json'",
+    "build:dev": "npm run build -- -- --source-maps inline",
+    "build:es": "tsc -p './build-tools/tsconfig-es.json'",
```
I think this is unneeded now as we're only shipping the commonjs to npm?

SzybkiSasza replied:
This still builds different code part that is accessible in the package via `webpack-filter-warnings-plugin/es`. CJS build generated by other pipeline is transpiled to be compatible with Node 4.x - so e.g. classes, spread operators etc. are replaced with compatible code, so it's essentially slower on newer NodeJS versions than `es` build.

mattlewis92 replied:
There aren't really any benefits to shipping an ES module version as well (today at least), as node still doesn't have native support for ES modules, so no one would be able to import this version without using a transpiler. Regarding the performance implications, it's such a small module that I doubt there would be any noticeable benefits to the end user. In fact, as node 6 reaches EOL next month, I think we should just drop support for it all together and require >= node 8 to use this package. As we're doing to bump the major anyway, I think that's acceptable. WDYT?

SzybkiSasza replied:
Yeah, I'm a bit of performance freak, so I got carried away :)

I completely agree, especially as babelified files contain flag that allows ES-compatible code to reuse module (`__esModule: true`), so it may be used without any problems in e.g. Typescript and other environments.

What if I adjust `.babelrc` and `package.json` so it targets at least Node 8 and drop `ES` config?

mattlewis92 replied:
Yup that sounds great! Can you also adjust the [node engines property](https://github.com/mattlewis92/webpack-filter-warnings-plugin/blob/master/package.json#L40) to be `>= 8.0.0`

mattlewis92 commented on package.json:
```diff
@@ -34,38 +42,42 @@
     "node": ">= 4.3 < 5.0.0 || >= 5.10"
   },
   "peerDependencies": {
+    "@types/webpack": "^4.4.25",
```
Let's remove this and stick with documenting it as required in the readme, if you're not using typescript then it's unfair to force users to install typings they don't need

SzybkiSasza replied:
As of npm@3 this does not get installed, it only triggers warning during installation

mattlewis92 replied:
People will still open issues complaining about it though, sadly npm doesn't have optional peer dependencies 😢

SzybkiSasza commented:
@mattlewis92 done! All the remarks are addressed, if it works for you, we should be good to go! :)

mattlewis92 commented:
Thanks again for your hard work on this! 😄 

Just published a beta to npm, would you mind installing it in your project and double checking everything is ok?

```
npm i webpack-filter-warnings-plugin@next
```
